File: 2022/day_04/rust/src/assignment.rs
type Section = u8;
pub type AssignmentPair = ((Section, Section), (Section, Section));

pub enum Overlap {
    None,
    Partial,
    Full,
}

/// Parses a given input file line and converts it into an assignment pair.
pub fn line_to_assignment_pair(line: &str) -> AssignmentPair {
    let mut parts: Vec<Section> = Vec::new();
    let assignments = line.split(',');
    for assignment in assignments {
        let sections = assignment.split('-');
        for section in sections {
            parts.push(section.parse::<Section>().unwrap());
        }
    }
    (
        (*parts.first().unwrap(), *parts.get(1).unwrap()),
        (*parts.get(2).unwrap(), *parts.get(3).unwrap()),
    )
}

/// Gets an overlap for a given assignment pair.
pub fn get_overlap(assignment_pair: &AssignmentPair) -> Overlap {
    let ((w, x), (y, z)) = assignment_pair;
    if (w <= y && x >= z) || (w >= y && x <= z) {
        Overlap::Full
    } else if (w <= y && y <= x) || (w <= z && z <= x) {
        Overlap::Partial
    } else {
        Overlap::None
    }
}

File: 2022/day_04/haskell/app/Assignment.hs
module Assignment where

import Data.List.Split (splitOn)

type Section = Int
type AssignmentPair = ((Section, Section), (Section, Section))

data Overlap = None | Partial | Full deriving (Eq)

-- Parses a given input file line and converts it into an assignment pair.
lineToAssignmentPair :: String -> AssignmentPair
lineToAssignmentPair s = ((head a1, a1 !! 1), (head a2, a2 !! 1))
  where
    stringByElf = splitOn [','] s
    a1 = stringByElfToAssignment 0
    a2 = stringByElfToAssignment 1
    stringByElfToAssignment i = map read $ splitOn ['-'] (stringByElf !! i)

-- Gets an overlap for a given assignment pair.
getOverlap :: AssignmentPair -> Overlap
getOverlap ((w, x), (y, z))
    | (w <= y && x >= z) || (w >= y && x <= z) = Full
    | (w <= y && y <= x) || (w <= z && z <= x) = Partial
    | otherwise = None
